### ISSUE TITLE
excludes in a plugin

### ISSUE THREAD

**Issue by iluuu1994** [lando/lando#2440]
We're writing a custom plugin to reduce boilerplate in our projects that all use the same configuration. We'd like to add [excludes](https://docs.lando.dev/config/performance.html#configuration) to that plugin because the directory structure of our projects is always the same.

I found out that when dumping the options in the plugin that the `excludes` is located in `options._app.config.excludes`.

```js
module.exports = {
    ...
    builder: (parent, config) => class LandoLamp extends parent {
        constructor(id, options = {}) {
            options = { ...config, ...options };
            options._app.config.excludes = [...];
            super(id, options);
        }
    },
};
```

However, setting that value in the plugin has no effect. The `_` in `_app` also gives me the impression that this is not something I should be accessing. Is there an official way to do this or is this currently not supported?

Thank you!

**Comment by dustinleblanc** [comment]
Hey @iluuu1994! Excludes is a top level property rather than a recipe property. The principal that it operates on is just creating named volumes within the docker container that get shared between the containers of the application without sharing those containers to the host. If you wanted to set this up within a recipe, you'd just create the volume mounts as part of the recipe rather than hooking into the top level excludes key.

Here is a quick search of the codebase for where excludes is used: https://github.com/lando/lando/search?q=excludes&unscoped_q=excludes

Directly using named volumes wouldn't give you the use of the mounter feature, but I believe if you want to add excludes programmatically, you actually want to modify and inject a new app object rather than alteration within a recipe.

I think the _app data is essentially a read only cache of the application so that you can pull data from it. 

I think if you want to modify the app object for a given recipe you want an `app.js` file in the root of your plugin that would operate similar to the Pantheon plugin here: https://github.com/lando/lando/blob/13f98bf50ae71b6f0cb494986222233e04a8aa6c/plugins/lando-pantheon/app.js#L13

**Comment by iluuu1994** [comment]
@dustinleblanc Thank you for your explanation, I'll try your suggestion and report back :)

**Comment by iluuu1994** [comment]
> I think if you want to modify the app object for a given recipe you want an `app.js` file in the root of your plugin that would operate similar to the Pantheon plugin here:

Unfortunately this doesn't seem to work either. The only existing plugin I could find that modifies `app.config` in `app.js` is `lando-recipes`.

https://github.com/lando/lando/blob/78f6344e68978a252b2c761a3dff7ac26d80fea2/plugins/lando-recipes/app.js#L34

Here's what I've tried:

```js
module.exports = (app, lando) => {
    if (app.config.recipe === '...') {
        app.events.on('pre-init', () => {
            const excludes = [...];

            app.config = lando.utils.merge(
                {excludes},
                app.config,
            );
        });
    }
};
```

The code is executed and the config looks exactly the same afterwards as when I set the values in `.lando.yml`. I also tried it with different event priorities and with no events at all, all with no effect.

I also briefly tried the custom mounter approach. The file `services/mounter/builder.js` will be picked up by lando. However it will still use the default mounter. The mounter is hard-coded in `land-sharing`. I wasn't sure if this could be customized without overriding that whole plugin as well.

https://github.com/lando/lando/blob/86ce983cdefcdd088f992b1213732e46a0428089/plugins/lando-sharing/app.js#L40

Sorry for the lengthy description. Since lando is a compiled binary it's quite hard for me to debug this.

Thanks for your response!
